core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java

```diff
@@ -26,42 +26,33 @@
 import org.elasticsearch.indices.query.IndicesQueriesRegistry;
 
 import java.io.IOException;
+import java.util.Objects;
 
 public class QueryParseContext {
 
     private static final ParseField CACHE = new ParseField("_cache").withAllDeprecated("Elasticsearch makes its own caching decisions");
     private static final ParseField CACHE_KEY = new ParseField("_cache_key").withAllDeprecated("Filters are always used as cache keys");
 
-    private XContentParser parser;
-    private ParseFieldMatcher parseFieldMatcher = ParseFieldMatcher.EMPTY;
+    private final XContentParser parser;
 
     private IndicesQueriesRegistry indicesQueriesRegistry;
 
-    public QueryParseContext(IndicesQueriesRegistry registry) {
-        this.indicesQueriesRegistry = registry;
+    public QueryParseContext(IndicesQueriesRegistry registry, XContentParser parser, ParseFieldMatcher parseFieldMatcher) {
```

Comment:
Comming back to this: the problem with the parseFieldMatcher stored inside AbstractXContentParser is that (currently on maste) we don't set it anywhere except in ObjectParserTests and when the parser is reset() on the QueryParseContext (where its broken btw.). I agree that it would be best if there was only one ParseFieldMatcher (either the one set on the parser, or the one in the QueryParseContext). 

Having PFM only in the parser is hard because:
* we'd ideally make parseFieldMatcher a ctor arg for all XContentParsers. This however is a huge change (tons of ways to create XContentParser, all would need to be changes), I'm not even sure if this would work because we would need the IndexSettings (where the flag for strict parsing comes from) is held
* if we keep the current get/setParseFieldMatcher on XContentParser it is hard to check that we alway set it correctly

On the other hand, keeping PFM only in the QueryShardContext requires us removing it from ObjectParser. I ran into many problems trying to do this (to me it seems the setters, getters where just introduced in 5f258b to make PFM play nicely with ObjectParser). 

So for now I'm a little bit stuck here. I'm leaning toward trying to removing the PFM from the parser and only store it in the QueryParseContext, but I'm struggling to remove it there because ObjectParser depends and so far I didn't find a way to get around the getter/setter on the parser.